File: MPC模型预测控制器-4-Octave完整案例讲解/code/octave/MPC_Test.m
%% MPC 模型预测控制示例（修复版）
% 来源：https://www.bilibili.com/opus/666629407425691664
% 兼容 GNU Octave 和 MATLAB。

clear;
close all;
clc;

% Octave 的 quadprog 位于 optim 包；MATLAB 不需要加载包。
if exist('OCTAVE_VERSION', 'builtin')
    try
        pkg load optim;
    catch
        warning(['未找到 Octave optim 包，将使用无约束二次规划的' ...
                 '线性方程回退解法。']);
    end
end

%% 第一步：定义离散状态空间模型 x(k+1) = A*x(k) + B*u(k)
A = [1, 0.1; -1, 2];
B = [0.2, 1; 0.5, 2];

n = size(A, 1);  % 状态数量
p = size(B, 2);  % 输入数量

assert(size(A, 2) == n, 'A 必须是方阵。');
assert(size(B, 1) == n, 'A 与 B 的行数必须一致。');

%% 第二步：定义 MPC 权重
Q = [100, 0; 0, 1];  % 预测区间内的状态权重
F = [100, 0; 0, 1];  % 终端状态权重
R = [1, 0; 0, 0.1];  % 控制输入权重

assert(isequal(size(Q), [n, n]), 'Q 的尺寸必须是 n x n。');
assert(isequal(size(F), [n, n]), 'F 的尺寸必须是 n x n。');
assert(isequal(size(R), [p, p]), 'R 的尺寸必须是 p x p。');

%% 第三步：设置仿真长度、初始状态和预测区间
k_steps = 100;
N = 5;

% k_steps 次状态更新会产生 k_steps+1 个状态点。
% 原帖只预分配了 k_steps 列，却在最后写入第 k_steps+1 列。
X_K = zeros(n, k_steps + 1);
U_K = zeros(p, k_steps);
X_K(:, 1) = [20; -20];

%% 第四步：构造二次规划矩阵
[E, H] = MPC_Matrices(A, B, Q, R, F, N);

%% 第五步：滚动优化并推进系统
for k = 1:k_steps
    U_K(:, k) = Prediction(X_K(:, k), E, H, N, p);
    X_K(:, k + 1) = A * X_K(:, k) + B * U_K(:, k);
end

%% 第六步：绘图
state_steps = 0:k_steps;
input_steps = 0:(k_steps - 1);
state_labels = arrayfun(@(i) sprintf('x%d', i), 1:n, ...
                            'UniformOutput', false);
input_labels = arrayfun(@(i) sprintf('u%d', i), 1:p, ...
                            'UniformOutput', false);

figure('Name', 'MPC simulation');

subplot(2, 1, 1);
plot(state_steps, X_K.', 'LineWidth', 1.2);
grid on;
xlabel('k');
ylabel('State');
title('State trajectories');
legend(state_labels, 'Location', 'best');

subplot(2, 1, 2);
plot(input_steps, U_K.', 'LineWidth', 1.2);
grid on;
xlabel('k');
ylabel('Input');
title('Control inputs');
legend(input_labels, 'Location', 'best');

File: MPC模型预测控制器-4-Octave完整案例讲解/code/python/mpc_test.py
"""MPC 模型预测控制完整案例的 Python 版本。"""

from __future__ import annotations

from pathlib import Path

import numpy as np

from mpc_matrices import mpc_matrices
from prediction import prediction


def simulate(steps: int = 100, horizon: int = 5) -> tuple[np.ndarray, np.ndarray]:
    """运行 MPC 闭环仿真，返回状态历史和输入历史。"""
    a = np.array([[1.0, 0.1], [-1.0, 2.0]])
    b = np.array([[0.2, 1.0], [0.5, 2.0]])

    q = np.array([[100.0, 0.0], [0.0, 1.0]])
    f = np.array([[100.0, 0.0], [0.0, 1.0]])
    r = np.array([[1.0, 0.0], [0.0, 0.1]])

    n = a.shape[0]
    p = b.shape[1]

    # steps 次状态更新会产生 steps+1 个状态点。
    states = np.zeros((n, steps + 1))
    inputs = np.zeros((p, steps))
    states[:, 0] = np.array([20.0, -20.0])

    e, h = mpc_matrices(a, b, q, r, f, horizon)

    for k in range(steps):
        inputs[:, k] = prediction(states[:, k], e, h, horizon, p)
        states[:, k + 1] = a @ states[:, k] + b @ inputs[:, k]

    return states, inputs


def save_results(states: np.ndarray, inputs: np.ndarray) -> None:
    """保存数值结果；若安装了 matplotlib，则同时保存曲线图。"""
    output_dir = Path(__file__).resolve().parent / "results"
    output_dir.mkdir(exist_ok=True)

    state_table = np.column_stack((np.arange(states.shape[1]), states.T))
    input_table = np.column_stack((np.arange(inputs.shape[1]), inputs.T))
    np.savetxt(
        output_dir / "states.csv",
        state_table,
        delimiter=",",
        header="k," + ",".join(f"x{i + 1}" for i in range(states.shape[0])),
        comments="",
    )
    np.savetxt(
        output_dir / "inputs.csv",
        input_table,
        delimiter=",",
        header="k," + ",".join(f"u{i + 1}" for i in range(inputs.shape[0])),
        comments="",
    )

    try:
        import matplotlib.pyplot as plt
    except ImportError:
        print("未安装 matplotlib：已保存 CSV，跳过绘图。")
        return

    figure, axes = plt.subplots(2, 1, figsize=(9, 7), constrained_layout=True)

    state_steps = np.arange(states.shape[1])
    for i, state in enumerate(states, start=1):
        axes[0].plot(state_steps, state, label=f"x{i}")
    axes[0].set(title="State trajectories", xlabel="k", ylabel="State")
    axes[0].grid(True)
    axes[0].legend()

    input_steps = np.arange(inputs.shape[1])
    for i, control_input in enumerate(inputs, start=1):
        axes[1].plot(input_steps, control_input, label=f"u{i}")
    axes[1].set(title="Control inputs", xlabel="k", ylabel="Input")
    axes[1].grid(True)
    axes[1].legend()

    figure.savefig(output_dir / "mpc_simulation.png", dpi=160)
    plt.close(figure)


def main() -> None:
    states, inputs = simulate()
    save_results(states, inputs)

    print("仿真完成")
    print(f"最终状态: {states[:, -1]}")
    print(f"各输入最大绝对值: {np.max(np.abs(inputs), axis=1)}")


if __name__ == "__main__":
    main()
